### ISSUE TITLE
cdk running in codebuild.   Does'nt not use --profile or Environment Variables.

### ISSUE THREAD

**Issue by mrpackethead** [aws/aws-cdk#14339]
cdk cli does not use profiles when running in codebuild.. 

I have tryed two methods,  
(a) using the --profile and giving it a profile which i created earlier.   The profiles are valid, because they work with the AWS cli,
(b) getting Accesskey/secrect/token using aws sts and exporting them to the environment..  Again these work with the aws cli.. 

In both cases, the cdk synth command just ignores the --profile and the environment,  and uses the codebuild credentials. Which means that it can't synth my stacks in the environments that they will be deployed. ( so lookups work ). 

Running these commands on the desktop works. ( both ways )..   Whats going on that makes it work differently in a codebuild container?

Here is an example of what i see in codebuild.

```
974 | [Container] 2021/04/23 07:03:25 Running command synth_role=$(aws sts assume-role --role-arn arn:aws:iam::6xxxxxxxxxxxx3:role/cdk-fudawl-lookup-role-6xxxxxxxxxxxx3-ap-southeast-2 --role-session-name lookup)
975 |  
976 | [Container] 2021/04/23 07:03:26 Running command export AWS_ACCESS_KEY_ID=$(echo $synth_role \| jq -r .Credentials.AccessKeyId)
977 |  
978 | [Container] 2021/04/23 07:03:26 Running command export AWS_SECRET_ACCESS_KEY=$(echo $synth_role \| jq -r .Credentials.SecretAccessKey)
979 |  
980 | [Container] 2021/04/23 07:03:26 Running command export AWS_SESSION_TOKEN=$(echo $synth_role \| jq -r .Credentials.SessionToken)
981 |  
982 | [Container] 2021/04/23 07:03:26 Running command cdk synth -a cdk.out/assembly-mutate-PlaceHolderStage2
983 | [Error at /mutate/PlaceHolderStage2/PlaceholderStack2] Need to perform AWS calls for account 6xxxxxxxxxx3, but the current credentials are for 2xxxxxxxxxxxx5
984 | Found errors
985 |  
986 | [Container] 2021/04/23 07:03:27 Command did not exit successfully cdk synth -a cdk.out/assembly-mutate-PlaceHolderStage2 exit status 1
987 | [Container] 2021/04/23 07:03:27 Phase complete: BUILD State: FAILED
```

  - **CDK CLI Version  :**
  - **Framework Version:**  1.100.0
  - **Node.js Version:** 12.18.3 
  
   **OS    Codebuild4.0           :**
  - **Language  cdk app is written in python3.8

--- 

This is :bug: Bug Report

**Comment by mrpackethead** [comment]
While further working on this problem i have discovered someting interesting.  It appears that cdk is partially using the credentials/role, but not fully.          I have attacehed the buildlog of my codebuild job ( that was generated using pipelines )

at line 2145 we  get the credentials for the cross account role.
starting at lines 2154, we export the credentials to the environment.  
Line 2160, confirms that we are going to use the role in from the 6xxxxxxxxx3 account. 
Line 2167 is where we synth the stack that will be deployed in the cross account.  ( Verbose Logging )
Line 2251 shows that the CDK_DEFAULT_ACCOUNT was set to 6xxxxxx3

Line 2273 is where the look up happens.   It is correct that some context information is missing, because there is ( deliberately ) no cdk.context.json ( at this stage ).      The Lookup is returning the correct results.   **_The only way the synth process ( running in the 2xxxxxxx5 account can do this is via the cdk-xxxcdkxx-lookup role ( that was provisioned by a bootstrap with an extra lookup role )._**   

Then at line 2306/07,  we get this error. 

"Need to perform AWS calls for account 6xxxxxxxxxx3, but the current credentials are for 2xxxxxxxxx5"

Huh?  CDK was using the credentials for account 6xxxxx3 just before, and it has not been changed..

Some thoughts;

(a) Running these commands manually at my desktop or on an EC2 machine runs just fine. 
(b) THe codebuild is running an Ubuntu based image..  My desktop is debian, and my EC2 is amazon linux2..  Could there be a difference in the way that Ubuntu is doing something.  
(c) Is there somethign about the codebuild image that is causing this to break.


[buildlog_scrubed.txt](https://github.com/aws/aws-cdk/files/6374472/buildlog_scrubed.txt)

**Comment by mrpackethead** [comment]
@skinny85 , i've opened a support case ( Case ID 8264301571 ) for this problem now...

**Comment by skinny85** [comment]
Hey @mrpackethead ,

thanks for opening the issue.

> (a) using the --profile and giving it a profile which i created earlier. The profiles are valid, because they work with the AWS cli,
> (b) getting Accesskey/secrect/token using aws sts and exporting them to the environment.. Again these work with the aws cli..

Can you show how do you do both of these in CodeBuild?

Thanks,
Adam

**Comment by mrpackethead** [comment]
I think the build spec is probably the most useful way. 

here is the way that I exported the environment vars.  A lot of whats in the prebuild is redundant, as it was left over from trying to use the profile as a command line switch. 

```
BuildSpec: 
          {
            "version": "0.2",
            "phases": {
              "pre_build": {
                "commands": [
                  "npm install -g aws-cdk",
                  "aws codeartifact login --tool pip --domain customername --repository CandIPyPI",
                  "pip install -r requirements.txt",
                  "env | sort",
                  "aws configure set role_arn arn:aws:iam::2xxxxxxxxx5:role/cdk-xxcdkxx-lookup-role-2xxxxxxxxx5-ap-southeast-2 --profile 2xxxxxxxxx5-ap-southeast-2",
                  "aws configure set region ap-southeast-2 --profile 2xxxxxxxxx5-ap-southeast-2",
                  "aws configure set credential_source EcsContainer --profile 2xxxxxxxxx5-ap-southeast-2",
                  "aws sts get-caller-identity --profile 2xxxxxxxxx5-ap-southeast-2",
                  "aws configure set role_arn arn:aws:iam::6xxxxxxxxxx3:role/cdk-xxcdkxx-lookup-role-6xxxxxxxxxx3-ap-southeast-2 --profile 6xxxxxxxxxx3-ap-southeast-2",
                  "aws configure set region ap-southeast-2 --profile 6xxxxxxxxxx3-ap-southeast-2",
                  "aws configure set credential_source EcsContainer --profile 6xxxxxxxxxx3-ap-southeast-2",
                  "aws sts get-caller-identity --profile 6xxxxxxxxxx3-ap-southeast-2",
                  "cat ~/.aws/config"
                ]
              },
              "build": {
                "commands": [
                  "cdk synth -v",
                  "synth_role=$(aws sts assume-role --role-arn arn:aws:iam::2xxxxxxxxx5:role/cdk-xxcdkxx-lookup-role-2xxxxxxxxx5-ap-southeast-2 --role-session-name lookup)",
                  "aws sts get-caller-identity",
                  "export AWS_ACCESS_KEY_ID=$(echo $synth_role | jq -r .Credentials.AccessKeyId)",
                  "export AWS_SECRET_ACCESS_KEY=$(echo $synth_role | jq -r .Credentials.SecretAccessKey)",
                  "export AWS_SESSION_TOKEN=$(echo $synth_role | jq -r .Credentials.SessionToken)",
                  "aws sts get-caller-identity",
                  "cdk synth -a cdk.out/assembly-mutate-PlaceHolderStage1 -v",
                  "aws sts get-caller-identity",
                  "unset AWS_ACCESS_KEY_ID",
                  "unset AWS_SECRET_ACCESS_KEY",
                  "unset AWS_SESSION_TOKEN",
                  "synth_role=$(aws sts assume-role --role-arn arn:aws:iam::6xxxxxxxxxx3:role/cdk-xxcdkxx-lookup-role-6xxxxxxxxxx3-ap-southeast-2 --role-session-name lookup)",
                  "aws sts get-caller-identity",
                  "export AWS_ACCESS_KEY_ID=$(echo $synth_role | jq -r .Credentials.AccessKeyId)",
                  "export AWS_SECRET_ACCESS_KEY=$(echo $synth_role | jq -r .Credentials.SecretAccessKey)",
                  "export AWS_SESSION_TOKEN=$(echo $synth_role | jq -r .Credentials.SessionToken)",
                  "aws sts get-caller-identity",
                  "cdk synth -a cdk.out/assembly-mutate-PlaceHolderStage2 -v",
                  "aws sts get-caller-identity",
                  "unset AWS_ACCESS_KEY_ID",
                  "unset AWS_SECRET_ACCESS_KEY",
                  "unset AWS_SESSION_TOKEN",
                  "cdk synth -v"
                ]
              }
            },
            "artifacts": {
              "base-directory": "cdk.out",
              "files": "**/*"
            }
          }
```          

Here is what i did trying to use profiles..

```
BuildSpec: |-
          {
            "version": "0.2",
            "phases": {
              "pre_build": {
                "commands": [
                  "npm install -g aws-cdk",
                  "aws codeartifact login --tool pip --domain customername --repository CandIPyPI",
                  "pip install -r requirements.txt",
                  "env | sort",
                  "aws configure set role_arn arn:aws:iam::2xxxxxxxxx5:role/cdk-xxcdkxx-lookup-role-2xxxxxxxxx5-ap-southeast-2 --profile 2xxxxxxxxx5-ap-southeast-2",
                  "aws configure set region ap-southeast-2 --profile 2xxxxxxxxx5-ap-southeast-2",
                  "aws configure set credential_source EcsContainer --profile 2xxxxxxxxx5-ap-southeast-2",
                  "aws sts get-caller-identity --profile 2xxxxxxxxx5-ap-southeast-2",
                  "aws configure set role_arn arn:aws:iam::6xxxxxxxxxx3:role/cdk-xxcdkxx-lookup-role-6xxxxxxxxxx3-ap-southeast-2 --profile 6xxxxxxxxxx3-ap-southeast-2",
                  "aws configure set region ap-southeast-2 --profile 6xxxxxxxxxx3-ap-southeast-2",
                  "aws configure set credential_source EcsContainer --profile 6xxxxxxxxxx3-ap-southeast-2",
                  "aws sts get-caller-identity --profile 6xxxxxxxxxx3-ap-southeast-2",
                  "cat ~/.aws/config"
                ]
              },
              "build": {
                "commands": [
                  "cdk synth -v",
                  "cdk synth -a cdk.out/assembly-mutate-PlaceHolderStage1 -v profile 2xxxxxxxxx5-ap-southeast-2",
                  "cdk synth -a cdk.out/assembly-mutate-PlaceHolderStage2 -v profile 6xxxxxxxxxx3-ap-southeast-2",
                  "cdk synth -v"
                ]
              }
            },
            "artifacts": {
              "base-directory": "cdk.out",
              "files": "**/*"
            }
          }
```          

The synth cmmands ( and the creation of profiles ) are generated dynamically by my cdk app.

**Comment by mrpackethead** [comment]
FYI, while i only provided the log for the build urnning the first method, the error was teh same for both.

**Comment by skinny85** [comment]
Hmm. This line is interesting:

```
                  "aws configure set credential_source EcsContainer --profile 2xxxxxxxxx5-ap-southeast-2",
```

Does a profile with that option set work outside CodeBuild with the CDK?

**Comment by mrpackethead** [comment]
I have not tested it outside of codebuild, because a credential source of EcsContainer is only meaningful inside the container. 

However this does work inside the container..  You'll see we use the profile just after we create it, and it returns the correct response.   ( check line 374 of the  log file )

**Comment by Gaurav2Github** [comment]
View your running Codebuild in Session Manager
https://docs.aws.amazon.com/codebuild/latest/userguide/session-manager.html
It may give you further insight what's happening

**Comment by skinny85** [comment]
@mrpackethead can you try running a simple JS SDK program in CodeBuild (can literally do `s3 ls`), and see if that picks up the ` credential_source EcsContainer`?

CDK uses the JS SDK for our auth, so I'm wondering whether the problem isn't in there.

**Comment by mrpackethead** [comment]
@skinny85 ... sure onto it, but i'm pretty sure thats going to work just fine, as it is using the profiles just fine to assume the roles.. 

line 1975: 
[Container] 2021/04/25 10:46:06 Running command synth_role=$(aws sts assume-role --role-arn arn:aws:iam::2xxxxxxxxx5:role/cdk-xxcdkxx-lookup-role-2xxxxxxxxx5-ap-southeast-2 --role-session-name lookup)


FYI.. I changed the Image from the default ubuntu to an AmazonLinux.. and that is just fined.

**Comment by skinny85** [comment]
Just to be clear - I meant `s3 ls` **using the JavaScript AWS SDK**.

Not the AWS CLI.

**Comment by mrpackethead** [comment]
[Container] 2021/04/26 23:03:27 Running command aws s3 ls --profile 6xxxxxxxx3-ap-southeast-2
--returns a sensible output--- its the list of s3 buckets  ( I just had to

31 | 2020-06-16 17:22:43 abcdedfgh
32 | 2020-06-16 17:29:38 xyzabcdefg

**Comment by mrpackethead** [comment]
I have no idea how to do that?  

> Just to be clear - I meant `s3 ls` **using the JavaScript AWS SDK**.
> 
> Not the AWS CLI.

**Comment by mrpackethead** [comment]
So, heres whats happened.

(a) adding some code to do teh s3 ls, with sdk, resulted in a stack that full ran.. it got my s3 buckets and then proceeded to finish the entire job.  That was a bit confusing..

(b) I removed the s3 ls, to see what difference it made.. the job still ran and finished. 

(c) I carefully diffed my work to see what was different. It was just  that I had installed aws-sdk..   ( npm install aws-sdk ) as a matter of course to do the s3 ls. 

(d) @skinny85  prompted a looking at sdk versions.. 

when cdk is installed in codebuild it was pulling aws-sdk 2.893.0 as a prerequiste.      installing aws-sdk was upgrading it to aws-sdk 2.894.    2.894 at the time of writing is only 6 hours old!     If i didtn know better someone had implmented a fix to deal with the credentials in ECS. 

In summary the fix for this problem is that upgrading sdk to 2.894 resolves the problem... It is very likely that a future upgrade of aws-cdk which will change its dependancys will fix the problem.   

Its suspected ( but not proved ) that the problem is related to how credentials using an EscContainer are dealt with.
